src/main/java/org/opentripplanner/middleware/triptracker/ManageLegTraversal.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    /**
+     * Get the expected traveler position using the current time and trip leg.
+     */
+    private static Segment getSegmentFromTime(Instant currentTime, Leg leg) {
+        List<Segment> segments = interpolatePoints(leg);
+        return getSegmentFromTime(leg.startTime.toInstant(), currentTime, segments);
+    }
+
+    /**
+     * Make sure that all the required leg parameters are present.
+     */
+    private static boolean canUseLeg(Leg leg) {
+        return
+            leg != null &&
+            leg.duration > 0 &&
+            leg.legGeometry != null &&
+            leg.legGeometry.points != null &&
+            !leg.legGeometry.points.isEmpty();
+    }
+
+    /**
+     * Get the expected leg by comparing the current time against the start and end time of each leg.
+     */
+    public static Leg getExpectedLeg(Instant timeNow, Itinerary itinerary) {
+        if (canUseTripLegs(itinerary)) {
+            for (int i = 0; i < itinerary.legs.size(); i++) {
+                Leg leg = itinerary.legs.get(i);
+                if (i == 0 &&
+                    leg.mode.equalsIgnoreCase("walk") &&
+                    leg.startTime != null &&
+                    timeNow.isBefore(leg.startTime.toInstant())) {
+                    // If the first leg is a walk leg and the traveler has decided to start the trip early.
+                    return leg;
+                }
+                if (leg.startTime != null &&
+                    leg.endTime != null &&
+                    isTimeInRange(
+                        leg.startTime.toInstant(),
+                        // Offset the end time by a faction to avoid exact times being attributed to the wrong leg.
+                        leg.endTime.toInstant().minus(1, ChronoUnit.MILLIS),
+                        timeNow
+                    )) {
+                    return leg;
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * A trip must have an itinerary that contains at least one leg to qualify for tracking.
+     */
+    private static boolean canUseTripLegs(Itinerary itinerary) {
+        return
+            itinerary != null &&
+            itinerary.legs != null &&
+            !itinerary.legs.isEmpty();
+    }
+
+    /**
+     * Using the duration of a leg and it's points, produce a list of segments each containing a representative
+     * coordinate and time spent in the segment.
+     */
+    public static List<Segment> interpolatePoints(Leg expectedLeg) {
+        List<Position> positions = PolylineUtils.decode(expectedLeg.legGeometry.points, 5);
+        double totalDistance = getDistanceTraversedForLeg(positions);
+        return (totalDistance > 0)
+            ? getTimeInSegments(positions, expectedLeg.duration / totalDistance, expectedLeg.mode)
+            : new ArrayList<>();
+    }
+
+    /**
+     * Get the segment which contains the current time.
+     */
+    public static Segment getSegmentFromTime(
```

Comment:
Not sure if its customary in this project, but in OTP we add `@Nullable` to methods that can return null.

Reply:
I'm not sure if it is, but I'm going to update to include.